cardano/src/block/verify_chain.rs

```diff
@@ -14,20 +14,34 @@ impl ChainState {
 
         add_error(&mut res, self.do_verify(block_hash, blk));
 
-        self.last_block = block_hash.clone();
-        self.last_date = Some(blk.header().blockdate());
-        // FIXME: count boundary blocks as part of the chain length?
-        self.chain_length += 1;
-
         match blk {
             Block::BoundaryBlock(blk) => {
                 self.last_boundary_block = Some(block_hash.clone());
+                self.last_boundary_block_epoch = Some(blk.header.consensus.epoch);
                 self.slot_leaders = blk.body.slot_leaders.clone();
             }
-
-            Block::MainBlock(_) => {}
+            Block::MainBlock(blk) => {
+                if self.last_boundary_block.is_some() {
+                    let block_epoch = blk.header.consensus.slot_id.epoch;
+                    let local_epoch = self.last_date.map(|d| d.get_epochid()).unwrap_or(0);
+                    if block_epoch > local_epoch {
+                        // We are in OBFT and switched epochs without EBB
+                        // Remove last boundary block state
+
+                        // TODO: should cleanup EBB state here?
+                        //self.last_boundary_block = None;
+                        //self.last_boundary_block_epoch = None;
+                        //self.slot_leaders = vec![];
```

Comment:
TODO: gotta decide whether we need to drop "Last EBB" state here after switch to OBFT. Cuz right now it just hangs there forever.

ℹ️ Gotta remember that "Last EBB" state is used to determine whether current epoch is an EBB or not, when storing epoch packs with meta-flags.